File: Java/Word_Ladder_II.java
public class Solution {
    
    private ArrayList<ArrayList<String>> ret = new ArrayList<ArrayList<String>>();
    private HashMap<String, HashSet<String>> predsMap =  new HashMap<String, HashSet<String>>();
    
    public ArrayList<ArrayList<String>> findLadders(String start, String end, HashSet<String> dict) {
        // Initialization
        ret.clear();
        predsMap.clear();
        for (String x : dict) {
            predsMap.put(x, new HashSet<String>());
        }
        predsMap.put(end, new HashSet<String>());
        bfsSearch(start, end, dict);
        ArrayList<String> cand = new ArrayList<String>();
        cand.add(end);
        genResults(cand, start, end);
        return ret;
    }
    
    private void bfsSearch(String start, String end, HashSet<String> dict) {
        HashSet<String> cands = new HashSet<String>();
        Queue<String> q = new LinkedList<String>();
        q.add(start);
        boolean endAtCurLevel = false;
        char original = 0;
        String tmp = null;
        while (!q.isEmpty()) {
            String word = q.remove();
            char wordArray[] = word.toCharArray();
            for (int i = 0; i < wordArray.length; i ++) {
                original = wordArray[i];
                for (char c = 'a'; c <= 'z'; c ++) {
                    wordArray[i] = c;
                    tmp = String.valueOf(wordArray);
                    if (end.equals(tmp)) {
                        endAtCurLevel = true;
                        predsMap.get(end).add(word);
                    }
                    if (dict.contains(tmp)) {
                        predsMap.get(tmp).add(word);
                        cands.add(tmp);
                    }
                }
                wordArray[i] = original;
            }
            if (q.isEmpty()) {
                if (endAtCurLevel) break;
                for (String x : cands) {
                    q.add(x); dict.remove(x);
                }
                cands.clear();
            }
        }
    }
    
    private void genResults(ArrayList<String> cand, String start, String end){
        if (end.equals(start)) {
            ret.add(cand);
            return;
        }
        HashSet<String> preds = predsMap.get(end);
        for (String x : preds) {
            ArrayList<String> newCand = new ArrayList<String>(cand);
            newCand.add(0, x);
            genResults(newCand, start, x);
        }
    }
}

File: cpp/Word_Ladder_II.cpp
class Solution {
public:
    vector<vector<string>> ret;
    unordered_map<string, unordered_set<string>> predsMap;
    vector<vector<string>> findLadders(string start, string end, unordered_set<string> &dict) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function

        // Clear return vector and the predecessor map
        ret.clear(); 
        predsMap.clear();

        // Initialize the predecessor map
        unordered_set<string>::iterator it;
        for (it = dict.begin(); it != dict.end(); ++it) {
            predsMap[*it] = unordered_set<string>();
        }
        predsMap[end] = unordered_set<string>();
       
        // Breadth first seach process
        queue<string> words;
        unordered_set<string> cands;
        words.push(start);
        int remain = 1;
        string tmp, cand;
        char original;
        bool endSearch = false;
        while (!words.empty()) {
            cand = words.front();
            words.pop();
            remain --;
            tmp = string(cand);
            for (int i = 0; i < tmp.length(); i ++) {
                original = tmp[i];    
                for (char c = 'a'; c <= 'z'; c++) {
                    tmp[i] = c;
                    if (tmp == end) {
                        endSearch = true;
                        predsMap[end].insert(cand);
                    }
                    if (!endSearch && dict.find(tmp) != dict.end()) {
                        cands.insert(tmp);
                        predsMap[tmp].insert(cand);
                    }
                }
                tmp[i] = original;
            }
            if (remain == 0) {          // ready to start the next level or terminate
                if (endSearch) break;
                for (it = cands.begin(); it != cands.end(); ++ it) {
                    words.push(*it);
                    dict.erase(*it);
                }
                remain = words.size();
                cands.clear();
            }
        }

        // generate the results
        genResults(vector<string>(1, end), start, end);
        return ret;

    }
   
    void genResults(vector<string> res, string start, string end){
        if (end == start) {
            reverse(res.begin(), res.end());
            ret.push_back(res);
            return;
        }
        unordered_set<string> preds = predsMap[end];
        unordered_set<string>:: iterator it;
        for (it = preds.begin(); it != preds.end(); ++it) {
            vector<string> newRes(res);
            newRes.push_back(*it);
            genResults(newRes, start, *it);
        }
    }
};
